Dashboard: Add tooltips for Twiss distribution parameters

proy30 opened the pull request:
The PR adds the tooltips for the twiss distribution parameters on the dashboard.

This is done by adding a helper function `get_tooltips_from_param`  in `Input/defaults_helper.py` which looks at a given function (in this case, `twiss` from`src\python\impactx\distribution_input_helpers.py`) and properly parses through the docstring by looking at each `:param` and its description to retrieve the tooltip.

proy30 committed:
change the method to retrieve docstring

now we bind the tooltip text dynamically via a vue state

allows more flexibility for tooltip lookups

proy30 committed:
Add tooltips for distribution parameters

Only for twiss

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
fix: codeQL

proy30 committed:
update: location of tooltip

set to 'top' so it doesn't block the error message on the bottom  or units on the right

EZoni commented on src/python/impactx/dashboard/Input/distributionParameters/distributionMain.py:
```diff
@@ -177,19 +177,27 @@ def card_content(self):
-                        )
+                        with vuetify.VTooltip(
+                            location="top",
+                            text=("all_tooltips[parameter.parameter_name]",),
```
I'm surprised that this works and you don't need 
```py
text=(f"all_tooltips[{parameter.parameter_name}]",),
```
instead. What am I missing?

EZoni commented on src/python/impactx/dashboard/Input/defaults_helper.py:
```diff
@@ -42,4 +45,26 @@ def get_docstrings(
+        doc = inspect.getdoc(func) or ""
+        pattern = re.compile(r"^\s*:param\s+(\w+)\s*:\s*(.+)$", re.MULTILINE)
+
+        for match in pattern.finditer(doc):
```
Do we maybe need to `try` this and raise an exception to avoid trying to call the method `finditer` from a `pattern` that may not have that method, e.g., in case the `re.compile` call that returned `pattern` did not run successfully (for some reason that we may or may not be able to anticipate)?

proy30 replied:
This is a good point. Right now we just assume the function we pass has the proper formatted docstring. We should add  a check to verify if that is not the case.

EZoni commented on src/python/impactx/dashboard/Input/defaults_helper.py:
```diff
@@ -42,4 +45,26 @@ def get_docstrings(
                     docstring = inspect.getdoc(attribute) or ""
                     docstrings[name] = docstring
 
+        distribution_tooltips = InputDefaultsHelper.get_tooltips_from_param(twiss)
```
Just double checking, this is intentionally parsing the Twiss docstring as the default docstring to be parsed?

proy30 replied:
For now, yes since Twiss is our default set of parameters. A future PR needs to add logic to retrieve the tooltips for when non-twiss parameters are selected, and in that PR this line of code possibly has to change.

cemitch99 reviewed (approved):
Thanks a lot.

proy30 committed:
add: guard clause for missing :param entries in docstring

the function to retrieve docstirngs from a function should always be non-empty to ensure the UI correctly displays the tooltip and ensure the function is properly documented

proy30 committed:
update: variable names and docstring
